File: app/src/main/java/com/omar/retromp3recorder/app/di/AudioPlayer.java
package com.omar.retromp3recorder.app.di;

import io.reactivex.Observable;

public interface AudioPlayer {
    void playerStop();

    void playerStart(String voiceURL);

    Observable<Event> observeEvents();

    interface Event {
    }

    class Message implements Event {
        public final String message;

        public Message(String errorMessage) {
            this.message = errorMessage;
        }
    }

    class Error extends Message {
        public Error(String errorMessage) {
            super(errorMessage);
        }
    }

    class SendPlayerId implements Event {
        public final int playerId;

        public SendPlayerId(int playerId) {
            this.playerId = playerId;
        }
    }

    class PlaybackEnded implements Event {
    }
}

File: app/src/main/java/com/omar/retromp3recorder/app/modules/playback/AudioPlayer.kt
package com.omar.retromp3recorder.app.modules.playback

import com.github.alkurop.stringerbell.Stringer
import io.reactivex.Observable

interface AudioPlayer {
    fun playerStop()
    fun playerStart(voiceURL: String)
    fun observeEvents(): Observable<Event>
    val isPlaying: Boolean

    //region events
    sealed class Event {
        data class Message(val message: Stringer) : Event()
        data class Error(val error: Stringer) : Event()
        data class PlayerId internal constructor(val playerId: Int) : Event()
    }

    fun observeState(): Observable<State>

    enum class State {
        Idle,
        Playing
    }
}